### PR TITLE
Replenish pool when release_conn=False and error occurred

### LINKED ISSUE #644: Connection pool exhausted when connection failures occur, should refill with empty connections
If preload_content is not specified on a request, connections are implicitly returned to the pool after the request is read. However, when a timeout error occurs or the connection is dropped the connection is closed but not returned in to the pool (since the response is None). This problem is compounded when retries are turned on since the next attempt will grab a new connection from the pool thus depleting it further. With non-blocking pools this is not a problem since we will create a new connection if the pool is empty but when the pool is blocking we have found that eventually the pool size becomes zero (after no.of connections timeout errors) which causes the calling application to hang on it's next request.
This can be fixed in connectionpool.urlopen in the exception handlers by explicitly returning the closed connection to the pool via _put_conn (if release_conn is false), since subsequent calls to _get_conn will check if the connection has been dropped and return a new one.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/20328594-connection-pool-exhausted-when-connection-failures-occur-should-refill-with-empty-connections?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by shazow**
Replaces #646
Fixes #644

### CODE REVIEW

**Comment by shazow** [comment]
@jlatherfold @Lukasa Please review.

Also need some more tests, I think socket-level. @Lukasa up for helping? <3 <3 <3

File: test/with_dummyserver/test_connectionpool.py [hunk 1]
```diff
@@ -647,6 +647,27 @@ def test_chunked_gzip(self):
 
         self.assertEqual(b'123' * 4, response.read())
 
+    def test_cleanup_on_connection_error(self):
```
Comment (shazow) [inline test/with_dummyserver/test_connectionpool.py]: Is this test even testing what we want to test?

**Comment by shazow** [comment]
Alright, I snuck in some extra assertions into existing tests, good enough! Please review, then merging.

File: test/with_dummyserver/test_connectionpool.py [hunk 2]
```diff
@@ -647,6 +647,27 @@ def test_chunked_gzip(self):
...
+        Test that connections are recycled to the pool on 
+        connection errors where no http response is received.
+        '''
+        poolsize = 3
+        with HTTPConnectionPool(self.host, self.port, maxsize=poolsize, block=True) as http:
+            self.assertEqual(http.pool.qsize(), poolsize)
+
+            # force a connection error by supplying a non-existent 
+            # url. We won't get a response for this  and so the 
+            # conn won't be implicitly returned to the pool.
```
Comment (Lukasa) [inline test/with_dummyserver/test_connectionpool.py]: s/won't/will/

File: urllib3/connectionpool.py [hunk 3]
```diff
@@ -409,7 +409,7 @@ def is_same_host(self, url):
 
         # TODO: Add optional support for socket.gethostbyname checking.
         scheme, host, port = get_host(url)
-
+
```
Comment (Lukasa) [inline urllib3/connectionpool.py]: =O Don't you have strip trailing whitespace in neovim?

**Comment by Lukasa** [comment]
Some notes, but otherwise :+1:

File: urllib3/connectionpool.py (hunk 3 above)
Comment (shazow) [inline urllib3/connectionpool.py]: I do but I like to run it manually (and forget).

File: test/with_dummyserver/test_connectionpool.py (hunk 2 above)
Comment (shazow) [inline test/with_dummyserver/test_connectionpool.py]: Fixed.

File: urllib3/connectionpool.py [hunk 4]
```diff
@@ -568,25 +568,22 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
...
         except (TimeoutError, HTTPException, SocketError, ConnectionError) as e:
-            if conn:
-                # Discard the connection for these exceptions. It will be
-                # be replaced during the next _get_conn() call.
-                conn.close()
-                conn = None
+            # Discard the connection for these exceptions. It will be
+            # be replaced during the next _get_conn() call.
+            conn = conn and conn.close()
+            release_conn = True
```
Comment (jlatherfold) [inline urllib3/connectionpool.py]: This could result in a read error on retry when we're not preloading content. release_conn should not be changed since it's used in subsequent recursive calls to this method. If you want to do this you should use a separate (local) variable. Retries should use the same value of release_conn as supplied to this call.

File: urllib3/connectionpool.py [hunk 5]
```diff
@@ -568,25 +568,22 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
...
             raise SSLError(e)
 
         except SSLError:
             # Treat SSLError separately from BaseSSLError to preserve
             # traceback.
-            if conn:
-                conn.close()
-                conn = None
+            conn = conn and conn.close()
+            release_conn = True
```
Comment (jlatherfold) [inline urllib3/connectionpool.py]: This could result in a read error on retry when we're not preloading content. release_conn should not be changed since it's used in subsequent recursive calls to this method. If you want to do this you should use a separate (local) variable. Retries should use the same value of release_conn as supplied to this call.

File: urllib3/connectionpool.py [hunk 6]
```diff
@@ -568,25 +568,22 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
             # Close the connection. If a connection is reused on which there
             # was a Certificate error, the next request will certainly raise
             # another Certificate error.
-            if conn:
-                conn.close()
-                conn = None
+            conn = conn and conn.close()
+            release_conn = True
```
Comment (jlatherfold) [inline urllib3/connectionpool.py]: This could result in a read error on retry when we're not preloading content. release_conn should not be changed since it's used in subsequent recursive calls to this method. If you want to do this you should use a separate (local) variable. Retries should use the same value of release_conn as supplied to this call.

Comment (shazow) [inline urllib3/connectionpool.py]: Hrm I could have sworn I left a reply but I guess it didn't take. —

How would the subsequent recursive call happen if we're raising an exception immediately after?

Comment (jlatherfold) [inline urllib3/connectionpool.py]: In the case of BaseSSLError, CertificateError or SSLError you raise right after setting release_conn to True but where is that exception caught? - I think that it will miss the finally block where you do that actual release with self._put_conn so it looks like those connections will never make it back to the pool.

In the TimeoutError, HTTPException, SocketError, ConnectionError exception handler you do not raise an exception. You set release_conn to True, release it in the finally block and then make a recursive call to urlopen, this time with release_conn set to True. If that succeeds you'll then release the conn back to the pool (because release_conn is now true) before the response is read, meaning that it's possible that when response.read() happens the connection might be in use by someone else.

Apologies if I'm wrong about this cos I haven't tested it yet but just by inspecting the code it looks that way.

File: urllib3/connectionpool.py (hunk 5 above)
Comment (jlatherfold) [inline urllib3/connectionpool.py]: Apologies, my first comment was incorrect but I think there is still a bug here triggered by the next line. If we raise from here then it looks like the connection will not be returned to the pool at all.

File: urllib3/connectionpool.py (hunk 6 above)
Comment (jlatherfold) [inline urllib3/connectionpool.py]: Sorry, I've got the lines mixed up.

The BaseSSLError, CertificateError, SSLError exception handlers raise after setting release_conn to true, but that will bypass the finally block and the the conn (or None in this case) is never put back on to the queue. For the (TimeoutError, HTTPException, SocketError, ConnectionError) exception handler you don't raise (presumably because there may be retries) and you will go on to release the conn and make another call to urlopen (if retries). But you call it with release_conn=True (since you've just set it to true) - that happens if that call succeeds? You'll release the conn back to the pool before the response is read. At least that's what it looks like to me.

File: urllib3/connectionpool.py (hunk 5 above)
Comment (shazow) [inline urllib3/connectionpool.py]: On L600:

```python
         finally:
             if release_conn:
                 # Put the connection back to be reused. If the connection is
                 # expired then it will be None, which will get replaced with a
                 # fresh connection during _get_conn.
                 self._put_conn(conn)
```

This will still get called just before the exception is raised.

File: urllib3/connectionpool.py (hunk 6 above)
Comment (shazow) [inline urllib3/connectionpool.py]: > The BaseSSLError, CertificateError, SSLError exception handlers raise after setting release_conn to true, but that will bypass the finally block and the the conn (or None in this case) is never put back on to the queue. 

Why will that bypass the finally block?

> For the (TimeoutError, HTTPException, SocketError, ConnectionError) exception handler you don't raise (presumably because there may be retries) and you will go on to release the conn and make another call to urlopen (if retries). But you call it with release_conn=True (since you've just set it to true) - what happens if that call succeeds?

Yea this condition is a little hard to follow, we should try to add a test for it. Hmm.

File: urllib3/connectionpool.py (hunk 5 above)
Comment (jlatherfold) [inline urllib3/connectionpool.py]: Your are right - not sure what I was thinking - of course the finally block will be executed.
